File: src/song/chord-builder.ts
import { type Chord, type Pitch, semitone } from "./model.ts";
import {
  shiftPitch,
  romanRoot,
  romanPitch,
  checkedPitch,
} from "./harmony-pitch.ts";
export const QUALITIES = [
  "major",
  "minor",
  "diminished",
  "augmented",
  "sus2",
  "sus4",
  "power",
] as const;
export interface ChordRecipe {
  root: string;
  quality: (typeof QUALITIES)[number];
  extension: 0 | 6 | 7 | 9 | 11 | 13;
  seventh: "major" | "minor" | "diminished";
  tones: { degree: number; alteration: number }[];
  omit: number[];
  inversion: number;
  octave: number;
  target: string | null;
  tonic: Pitch;
}
export function buildChord(id: string, name: string, r: ChordRecipe): Chord {
  if (
    !r ||
    !QUALITIES.includes(r.quality) ||
    ![0, 6, 7, 9, 11, 13].includes(r.extension) ||
    !["major", "minor", "diminished"].includes(r.seventh)
  )
    throw new Error("Choose chord quality, extension and seventh quality");
  if (
    r.extension >= 7 &&
    r.seventh === "diminished" &&
    r.quality !== "diminished"
  )
    throw new Error(
      "Diminished sevenths require diminished quality; use an added sixth for other qualities",
    );
  checkedPitch(r.tonic);
  if (
    !Array.isArray(r.tones) ||
    r.tones.length > 13 ||
    !Array.isArray(r.omit) ||
    r.omit.length > 13 ||
    new Set(r.omit).size !== r.omit.length ||
    !r.omit.every((n) => Number.isInteger(n) && n >= 1 && n <= 13) ||
    !Number.isInteger(r.octave) ||
    Math.abs(r.octave) > 4
  )
    throw new Error("Invalid chord tones, omissions or octave");
  const root = romanRoot(r.root),
    target = r.target === null ? null : romanRoot(r.target);
  const base = target
    ? shiftPitch(r.tonic, target.degree - 1, semitone(target))
    : r.tonic;
  const pitch = shiftPitch(
    base,
    root.degree - 1 + 7 * r.octave,
    semitone(root) + 12 * r.octave,
  );
  const thirds = {
    major: 4,
    minor: 3,
    diminished: 3,
    augmented: 4,
    sus2: 2,
    sus4: 5,
    power: 0,
  };
  const toneMap = new Map<number, number>([
    [1, 0],
    [5, r.quality === "diminished" ? 6 : r.quality === "augmented" ? 8 : 7],
  ]);
  if (r.quality !== "power")
    toneMap.set(
      r.quality === "sus2" ? 2 : r.quality === "sus4" ? 4 : 3,
      thirds[r.quality],
    );
  if (r.extension === 6) toneMap.set(6, 9);
  if (r.extension >= 7)
    toneMap.set(7, r.seventh === "major" ? 11 : r.seventh === "minor" ? 10 : 9);
  if (r.extension >= 9) toneMap.set(9, 14);
  if (r.extension >= 11) toneMap.set(11, 17);
  if (r.extension >= 13) toneMap.set(13, 21);
  const natural = [0, 2, 4, 5, 7, 9, 11];
  const modified = new Set<number>();
  for (const t of r.tones) {
    if (
      !t ||
      !Number.isInteger(t.degree) ||
      t.degree < 1 ||
      t.degree > 13 ||
      !Number.isInteger(t.alteration) ||
      Math.abs(t.alteration) > 2 ||
      modified.has(t.degree)
    )
      throw new Error(
        "Added/altered tones need unique degrees 1–13 and alterations -2…2",
      );
    modified.add(t.degree);
    toneMap.set(
      t.degree,
      natural[(t.degree - 1) % 7]! +
        12 * Math.floor((t.degree - 1) / 7) +
        t.alteration,
    );
  }
  for (const d of r.omit) toneMap.delete(d);
  if (!toneMap.size) throw new Error("A chord must retain at least one note");
  const notes = [...toneMap]
    .map(([degree, n]) => ({
      id: `tone-${degree}`,
      pitch: shiftPitch(pitch, degree - 1, n),
    }))
    .sort((a, b) => semitone(a.pitch) - semitone(b.pitch));
  if (
    !Number.isInteger(r.inversion) ||
    r.inversion < 0 ||
    r.inversion >= notes.length
  )
    throw new Error("Inversion must select a retained chord tone");
  for (let i = 0; i < r.inversion; i++) {
    const n = notes.shift()!;
    do {
      n.pitch = shiftPitch(n.pitch, 7, 12);
    } while (semitone(n.pitch) <= semitone(notes.at(-1)!.pitch));
    notes.push(n);
  }
  let suffix =
    r.quality === "diminished"
      ? r.extension >= 7 && r.seventh === "minor"
        ? "ø"
        : "°"
      : r.quality === "augmented"
        ? "+"
        : r.quality === "power" && !r.extension
          ? "5"
          : "";
  if (r.extension)
    suffix +=
      (r.extension >= 7 && r.seventh === "major" ? "maj" : "") + r.extension;
  if (r.quality.startsWith("sus")) suffix += r.quality;
  const details = [
    ...r.tones.map(
      (t) =>
        `${t.alteration < 0 ? "♭".repeat(-t.alteration) : "♯".repeat(t.alteration)}${t.degree}`,
    ),
    ...r.omit.map((n) => `no${n}`),
    ...(r.quality === "power" &&
    r.extension &&
    !toneMap.has(3) &&
    !r.omit.includes(3)
      ? ["no3"]
      : []),
    ...(r.inversion ? [`bass ${notes[0]!.id.slice(5)}`] : []),
  ];
  return {
    id,
    name,
    notes,
    labelTonic: { ...r.tonic },
    label: `${romanPitch(root, r.quality === "minor" || r.quality === "diminished")}${suffix}${r.target ? `/${r.target}` : ""}${details.length ? `(${details.join(",")})` : ""}`,
  };
}

File: src-tauri/crates/song-core/src/chord_builder.rs
//! Chord construction from Roman-root recipes. Mirrors chord-builder.ts.
use crate::{
    Pitch,
    harmony_pitch::{checked_pitch, roman_pitch, roman_root, semitone, shift_pitch},
};
use crate::{Result, ensure};
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

pub const QUALITIES: [&str; 7] = [
    "major",
    "minor",
    "diminished",
    "augmented",
    "sus2",
    "sus4",
    "power",
];

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ts_rs::TS)]
#[serde(rename_all = "camelCase")]
pub struct RecipeTone {
    pub degree: i32,
    pub alteration: i32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ts_rs::TS)]
#[serde(rename_all = "camelCase")]
pub struct ChordRecipe {
    pub root: String,
    pub quality: String,
    pub extension: i32,
    pub seventh: String,
    pub tones: Vec<RecipeTone>,
    pub omit: Vec<i32>,
    pub inversion: i32,
    pub octave: i32,
    pub target: Option<String>,
    #[ts(type = "{ degree: number, alteration: number, octave: number }")]
    pub tonic: Pitch,
}

fn integer(value: i32, lo: i32, hi: i32) -> bool {
    (lo..=hi).contains(&value)
}

pub fn build_chord(id: String, name: String, recipe: &ChordRecipe) -> Result<crate::Chord> {
    ensure(
        QUALITIES.contains(&recipe.quality.as_str())
            && [0, 6, 7, 9, 11, 13].contains(&recipe.extension)
            && ["major", "minor", "diminished"].contains(&recipe.seventh.as_str()),
        "Choose chord quality, extension and seventh quality",
    )?;
    ensure(
        !(recipe.extension >= 7
            && recipe.seventh == "diminished"
            && recipe.quality != "diminished"),
        "Diminished sevenths require diminished quality; use an added sixth for other qualities",
    )?;
    checked_pitch(&recipe.tonic)?;
    ensure(
        recipe.tones.len() <= 13
            && recipe.omit.len() <= 13
            && recipe
                .omit
                .iter()
                .collect::<std::collections::BTreeSet<_>>()
                .len()
                == recipe.omit.len()
            && recipe.omit.iter().all(|n| integer(*n, 1, 13))
            && recipe.octave.abs() <= 4,
        "Invalid chord tones, omissions or octave",
    )?;
    let root = roman_root(&recipe.root)?;
    let target = recipe.target.as_ref().map(|t| roman_root(t)).transpose()?;
    let base = match &target {
        Some(target) => shift_pitch(
            &recipe.tonic,
            i64::from(target.degree - 1),
            i64::from(semitone(target)),
        )?,
        None => recipe.tonic.clone(),
    };
    let pitch = shift_pitch(
        &base,
        i64::from(root.degree - 1 + 7 * recipe.octave),
        i64::from(semitone(&root)) + 12 * i64::from(recipe.octave),
    )?;
    let thirds = match recipe.quality.as_str() {
        "major" => 4,
        "minor" => 3,
        "diminished" => 3,
        "augmented" => 4,
        "sus2" => 2,
        "sus4" => 5,
        _ => 0,
    };
    let mut tone_map = BTreeMap::from([
        (1, 0),
        (
            5,
            if recipe.quality == "diminished" {
                6
            } else if recipe.quality == "augmented" {
                8
            } else {
                7
            },
        ),
    ]);
    if recipe.quality != "power" {
        tone_map.insert(
            if recipe.quality == "sus2" {
                2
            } else if recipe.quality == "sus4" {
                4
            } else {
                3
            },
            thirds,
        );
    }
    if recipe.extension == 6 {
        tone_map.insert(6, 9);
    }
    if recipe.extension >= 7 {
        tone_map.insert(
            7,
            if recipe.seventh == "major" {
                11
            } else if recipe.seventh == "minor" {
                10
            } else {
                9
            },
        );
    }
    if recipe.extension >= 9 {
        tone_map.insert(9, 14);
    }
    if recipe.extension >= 11 {
        tone_map.insert(11, 17);
    }
    if recipe.extension >= 13 {
        tone_map.insert(13, 21);
    }
    let natural = [0, 2, 4, 5, 7, 9, 11];
    let mut modified = std::collections::BTreeSet::new();
    for tone in &recipe.tones {
        ensure(
            (1..=13).contains(&tone.degree)
                && (-2..=2).contains(&tone.alteration)
                && modified.insert(tone.degree),
            "Added/altered tones need unique degrees 1–13 and alterations -2…2",
        )?;
        tone_map.insert(
            tone.degree,
            natural[((tone.degree - 1) % 7) as usize]
                + 12 * ((tone.degree - 1) / 7)
                + tone.alteration,
        );
    }
    for degree in &recipe.omit {
        tone_map.remove(degree);
    }
    ensure(
        !tone_map.is_empty(),
        "A chord must retain at least one note",
    )?;
    let has_third = tone_map.contains_key(&3);
    // Sorted by sounding pitch; ties keep degree order on both sides.
    let mut notes: Vec<(i32, crate::Note)> = tone_map
        .into_iter()
        .map(|(degree, n)| {
            shift_pitch(&pitch, i64::from(degree - 1), i64::from(n)).map(|pitch| {
                (
                    semitone(&pitch),
                    crate::Note {
                        id: format!("tone-{degree}"),
                        pitch,
                    },
                )
            })
        })
        .collect::<Result<Vec<_>>>()?;
    notes.sort_by_key(|(semitones, _)| *semitones);
    let mut notes: Vec<crate::Note> = notes.into_iter().map(|(_, note)| note).collect();
    ensure(
        recipe.inversion >= 0 && (recipe.inversion as usize) < notes.len(),
        "Inversion must select a retained chord tone",
    )?;
    for _ in 0..recipe.inversion {
        let mut note = notes.remove(0);
        loop {
            note.pitch = shift_pitch(&note.pitch, 7, 12)?;
            let last = semitone(&notes.last().expect("retained note").pitch);
            if semitone(&note.pitch) > last {
                break;
            }
        }
        notes.push(note);
    }
    let mut suffix = if recipe.quality == "diminished" {
        if recipe.extension >= 7 && recipe.seventh == "minor" {
            "ø"
        } else {
            "°"
        }
    } else if recipe.quality == "augmented" {
        "+"
    } else if recipe.quality == "power" && recipe.extension == 0 {
        "5"
    } else {
        ""
    }
    .to_string();
    if recipe.extension != 0 {
        suffix += &format!(
            "{}{}",
            if recipe.extension >= 7 && recipe.seventh == "major" {
                "maj"
            } else {
                ""
            },
            recipe.extension
        );
    }
    if recipe.quality.starts_with("sus") {
        suffix += &recipe.quality;
    }
    let mut details: Vec<String> = recipe
        .tones
        .iter()
        .map(|t| {
            let accidental = if t.alteration < 0 {
                "♭".repeat((-t.alteration) as usize)
            } else {
                "♯".repeat(t.alteration as usize)
            };
            format!("{accidental}{}", t.degree)
        })
        .chain(recipe.omit.iter().map(|n| format!("no{n}")))
        .collect();
    if recipe.quality == "power" && recipe.extension != 0 && !has_third && !recipe.omit.contains(&3)
    {
        details.push("no3".into());
    }
    if recipe.inversion != 0 {
        details.push(format!(
            "bass {}",
            notes[0].id.strip_prefix("tone-").unwrap_or(&notes[0].id)
        ));
    }
    Ok(crate::Chord {
        id,
        name,
        notes,
        label_tonic: recipe.tonic.clone(),
        label: Some(format!(
            "{}{}{}{}",
            roman_pitch(
                &root,
                recipe.quality == "minor" || recipe.quality == "diminished"
            ),
            suffix,
            recipe
                .target
                .as_ref()
                .map(|t| format!("/{t}"))
                .unwrap_or_default(),
            if details.is_empty() {
                String::new()
            } else {
                format!("({})", details.join(","))
            }
        )),
    })
}
